### PR TITLE
Add guide for writing LSIF indexers

### PR DESCRIPTION
**Pull request by efritz**
This closes https://github.com/sourcegraph/sourcegraph/issues/9719.

### CODE REVIEW

**Comment by slimsag** [comment]
oh god haha, please lets not start committing docs images to this repository too.. we really need some check in place to prevent this. At least with github.com/sourcegraph/about it's fine if we have to blow away the entire history a year or two from now haha, but here that is so, so much worse

Can you upload these images to a GCP bucket instead?

**Comment by slimsag** [comment]
Here is a bucket folder you can put them in: https://console.cloud.google.com/storage/browser/sourcegraph-assets/docs.sourcegraph.com/?forceOnBucketsSortingFiltering=false&authuser=1&project=sourcegraph-dev

File: doc/user/code_intelligence/writing_an_indexer.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+{"id": "100", "type": "vertex", "label": "moniker", "kind": "import", "scheme": "gomod", "identifier": "github.com/slimsag/godocmd:ToMarkdown"}
+{"id": "101", "type": "vertex", "label": "packageInformation", "name": "github.com/slimsag/godocmd", "manager": "gomod", "version": "v0.0.0-a1005ad29fe3"}
+{"id": "102", "type": "edge", "label": "nextMoniker", "outV": "2", "inV": "100"}
+{"id": "103", "type": "edge", "label": "packageInformation", "outV": "100", "inV": "101"}
+```
+
+<center><img src="https://storage.googleapis.com/sourcegraph-assets/docs.sourcegraph.com/lsif-graphviz/7.png" width="33%"></center>
+
+<!--
+Here is the dot file used to generate the SVG images in this article (rendered via https://dreampuf.github.io/GraphvizOnline/).
```
Comment (gbrik) [inline doc/user/code_intelligence/writing_an_indexer.md:201]: suggestion:

expanding on the "more info" comment at the top, I'd specifically like these graphs to be detailed enough to serve as a reference once you've read the documentation. I have lots of _opinions_ on diagrams so I'd be happy to do them if you want (e.g. graphviz layouts make me :( )

Comment (efritz) [reply doc/user/code_intelligence/writing_an_indexer.md:201]: Yes, if you want to take a whack at updating the diagrams here PLEASE DO!

File: doc/user/code_intelligence/writing_an_indexer.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+{"id": "18", "type": "vertex", "label": "referenceResult"}
+{"id": "19", "type": "edge", "label": "textDocument/references", "outV": "11", "inV": "18"}
+```
+
+Finally, we fill out the relationships in the opposite direction. The definition result needs to link back to the range that defines it via an _item_ edge, and the reference result needs to link back to the set of ranges that define the symbol, and the set of ranges that reference the symbol via _item_ edges tagged with a _definitions_ and _references_ property, respectively.
+
+```json
+{"id": "20", "type": "edge", "label": "item", "outV": "16", "inVs": ["6"], "document": "3"}
+{"id": "21", "type": "edge", "label": "item", "outV": "18", "inVs": ["6"], "document": "3", "property": "definitions"}
+{"id": "22", "type": "edge", "label": "item", "outV": "18", "inVs": ["5"], "document": "3", "property": "references"}
```
[the remark this answers is not in the document]
Comment (efritz) [reply doc/user/code_intelligence/writing_an_indexer.md:128]: Not my spec!

[the remark this answers is not in the document]
Comment (efritz) [reply doc/user/code_intelligence/writing_an_indexer.md:128]: The first one is from a reference result to a range, which doesn't have a property. I think this is correct

(however the diagram mislabels 18 as 19, so there is at least one error here)

File: doc/user/code_intelligence/writing_an_indexer.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
+# Writing an indexer
+
+This page is a deep-dive into [LSIF](https://github.com/Microsoft/language-server-protocol/blob/master/indexFormat/specification.md) (Language Server Index Format) and is meant to be a general guide to writing an LSIF indexer for a language that isn't already supported on [LSIF.dev](https://lsif.dev). If you are interested in adding your LSIF indexer to the list, create a pull request [here](https://github.com/lsif/lsif.github.io). Sourcegraph currently reads versions 0.4.0 and 0.4.3 of the LSIF specification, but will continue to add the ability to read newer versions of the specification as they are published. Reference [Microsoft's documentation](https://microsoft.github.io/language-server-protocol/specifications/lsif/0.4.0/specification/) as an additional resource.
+
+## LSIF
```
[the remark this answers is not in the document]
Comment (efritz) [reply doc/user/code_intelligence/writing_an_indexer.md:5]: What did you have in mind here for content?

**Comment by abitrolly** [comment]
Available here https://docs.sourcegraph.com/code_intelligence/explanations/writing_an_indexer

**Pull request merged**